src/middlewared/middlewared/plugins/zfs.py

```diff
@@ -1431,3 +1434,49 @@ def rollback_impl(self, args, id):
             )
         except subprocess.CalledProcessError as e:
             raise CallError(f'Failed to rollback snapshot: {e.stderr.strip()}')
+
+    @accepts(
+        Str('id'),
+        Dict(
+            'options',
+            Bool('recursive', default=False),
+        ),
+    )
+    def hold(self, id, options):
+        """
+        Holds snapshot `id`.
+
+        `truenas` tag will be added to the snapshot's tag namespace.
+
+        `options.recursive` will hold snapshots recursively.
+        """
+        try:
+            with libzfs.ZFS() as zfs:
+                snapshot = zfs.get_snapshot(id)
+                snapshot.hold('truenas', options['recursive'])
+        except libzfs.ZFSException as err:
+            raise CallError(f'Failed to hold snapshot: {err}')
+
+    @accepts(
+        Str('id'),
+        Dict(
+            'options',
+            Bool('recursive', default=False),
+        ),
+    )
+    def release(self, id, options):
```

Comment:
This is a public method, let's please add returns decorator as well

Reply:
Isn't ZFS plugin private?